wire-ios-data-model/Source/MLS/Migration/ProteusToMLSMigrationCoordinator.swift

```diff
@@ -0,0 +1,297 @@
+
+    private typealias GroupIDConversationTuple = (groupID: MLSGroupID, conversation: ZMConversation)
+
+    private func fetchMixedConversations() async throws -> [GroupIDConversationTuple] {
+        return try await context.perform { [self] in
+
+            let conversations = try ZMConversation.fetchAllTeamGroupConversations(
+                messageProtocol: .mixed,
+                in: context
+            )
+
+            let tuples: [(MLSGroupID, ZMConversation)] = conversations.compactMap {
+                guard let groupID = $0.mlsGroupID else {
+                    return nil
+                }
+                return (groupID: groupID, conversation: $0)
+            }
+
+            return tuples
+        }
+    }
+
+    private func joinMLSGroupIfNeeded(_ groupID: MLSGroupID, mlsService: MLSServiceInterface) async throws {
+        if await mlsService.conversationExists(groupID: groupID) {
+            return
+        }
+
+        try await mlsService.joinGroup(with: groupID)
+    }
+
+    private func allParticipantsSupportMLS(in conversation: ZMConversation) -> Bool {
+        conversation.localParticipants
+            .allSatisfy { $0.supportedProtocols.contains(.mls) }
+    }
+
+    private func syncUsersWithTheBackend() async throws {
+        let fetchRequest = ZMUser.fetchRequest()
+
+        let qualifiedIDs = try await context.perform { [context] in
+            let users = try context.fetch(fetchRequest) as? [ZMUser]
+            return users?.compactMap { $0.qualifiedID }
+        }
+
+        guard let qualifiedIDs else { return }
+
+        try await actionsProvider.syncUsers(qualifiedIDs: qualifiedIDs, context: context.notificationContext)
+    }
+
+    private func migrationFinalisationTimeHasArrived() async -> Bool {
+        let mlsMigrationFeature = await context.perform { [featureRepository] in
+            featureRepository.fetchMLSMigration()
+        }
+
+        guard let finaliseDate = mlsMigrationFeature.config.finaliseRegardlessAfter else {
+            return false
+        }
+
+        return finaliseDate.isPast
+    }
+
+    private func updateConversationProtocolToMLS(for conversation: ZMConversation) async throws {
+        let qualifiedID = await context.perform { conversation.qualifiedID }
+        guard let qualifiedID else { return }
+
+        let messageProtocol: MessageProtocol = .mls
+
+        try await actionsProvider.updateConversationProtocol(
+            qualifiedID: qualifiedID,
+            messageProtocol: messageProtocol,
+            context: context.notificationContext
+        )
+
+        try await postProtocolChangeUpdater.updateLocalConversation(
+            for: qualifiedID,
+            to: messageProtocol,
+            context: context
+        )
+    }
+}
+
+// This is temporary until John's work on 1on1 conversations
+// is merged to develop and then we can pull the changes into the migration branch.
+// TODO: [AGIS] Get rid of this extension
```

Comment:
There are a few TODOs regarding the protocol updates, but I assume this will be solved with a future task when everything is merged together. So it's no blocker.

Reply:
**suggestion**" can you add the jira ticket where it should be resolved?

Reply:
Agis and me didn't find the ticket right away in this case. But good suggestion and let's follow this in future!